derive/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+extern crate arbitrary;
+extern crate syn;
+#[macro_use]
+extern crate synstructure;
+extern crate proc_macro2;
+
+use proc_macro2::TokenStream;
+
+decl_derive!([Arbitrary] => arbitrary_derive);
+
+fn arbitrary_derive(s: synstructure::Structure) -> TokenStream {
+    if s.variants().len() == 1 { // struct
+        let con = s.variants()[0].construct(|_, _| quote! { Arbitrary::arbitrary(u)? });
+        s.gen_impl(quote! {
+            extern crate arbitrary;
+
+            use arbitrary::{Arbitrary, Unstructured};
+
+            gen impl Arbitrary for @Self {
+                fn arbitrary<U: Unstructured + ?Sized>(u: &mut U) -> Result<Self, U::Error> {
+                    Ok(#con)
+                }
+            }
+        })
+    } else { // enum
+        let mut variant_tokens = TokenStream::new();
+
+        for (count, variant) in s.variants().iter().enumerate() {
+            let count = count as u64;
+            let constructor = variant.construct(|_, _| quote! { Arbitrary::arbitrary(u)? });
+            variant_tokens.extend(quote! { #count => #constructor, });
+        }
+        let count = s.variants().len() as u64;
+        s.gen_impl(quote! {
+            extern crate arbitrary;
+
+            use arbitrary::{Arbitrary, Unstructured};
+
+            gen impl Arbitrary for @Self {
+                fn arbitrary<U: Unstructured + ?Sized>(u: &mut U) -> Result<Self, U::Error> {
+
+
+                    Ok(match (u64::from(<u32 as Arbitrary>::arbitrary(u)?) * #count) >> 32 {
```

Comment:
I’m having a hard time understanding what this means. We generate u32, convert it to u64, multiply by the number of variants and then drop the low bytes? What is the purpose of this complex logic? Why isn’t the genereated number taken modulo variant count?

Perhaps a comment would be appropriate.

Reply:
This version introduces less bias than modulo and is due to [lemire](https://lemire.me/blog/2016/06/30/fast-random-shuffling/).

Reply:
Please extend the comment with this blog post and also mention that this has a very slight bias.